File: txtscrio/show_cstr.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#define LOOP_HEAD                                                            \
  unsigned char c; /* character to show */                                   \
  uint_least8_t flag; /* 0=read, 1=stop, 2=show */                           \
                                                                             \
  do {                                                                       \
    c = *s;                                                                  \
    s++;                                                                     \
                                                                             \
    if (c == '~') { /* attribute switch mark */                              \
      uint_least16_t t = a2;                                                 \
      a2 = a1;                                                               \
      a1 = t;                                                                \
      flag = (i == n) & 1; /* check string end */                            \
      n--;                                                                   \
    } else                                                                   \
      flag = 2;                                                              \
                                                                             \
  } while (!flag);                                                           \
                                                                             \
  if (flag == 1)                                                             \
    break;

#define LOOP_TAIL                                                            \
  if (i == n) /* check string end */                                         \
    break;                                                                   \
                                                                             \
  p++;                                                                       \
  i++;

#define LOOP_TAIL_X                                                          \
  LOOP_TAIL                                                                  \
  x++;

void show_cstr (uint8_t xpos, uint8_t ypos, const String *str, uint8_t attr1,
                uint8_t attr2) {

  uint_least8_t n; // number of input characters left to proceed

  n = Length (str);
  if (n) {
    const unsigned char *s; // input string
    uint16_t *p; // output screen address
    uint_least16_t a1, a2; // prepared attributes
    uint_least8_t i; // index for a new character to show (1 based)
    uint_least8_t x, y; // character's screen coordinates

    s = GetString (str);
    a1 = attr1 << 8;
    a2 = attr2 << 8;
    x = xpos;
    y = ypos;
    p = ((uint16_t *) screen_ptr) + (ypos-1)*MaxCol + xpos-1;
    i = 1;

    if ((y < area_y1) || (y > area_y2+1)) {
      while (1) {
        LOOP_HEAD

        *p = c | a1; // normal

        LOOP_TAIL
      }
    } else if (y == area_y1) {
      while (1) {
        LOOP_HEAD

        if ((x < area_x1) || (x > area_x2))
          *p = c | a1; // normal

        LOOP_TAIL_X
      }
    } else if (y <= area_y2) {
      uint_least8_t x22 = area_x2+2; // shadow: vertical rectangle's end X

      while (1) {
        LOOP_HEAD

        if ((x < area_x1) || (x > x22))
          *p = c | a1; // normal
        else if (x > area_x2)
          *((unsigned char *) p) = c; // shadow

        LOOP_TAIL_X
      }
    } else {
      uint_least8_t x12 = area_x1+2; // shadow: bottom line's start X
      uint_least8_t x22 = area_x2+2; // shadow: bottom line's end X

      while (1) {
        LOOP_HEAD

        if ((x < x12) && (x > x22))
          *p = c | a1; // normal
        else
          *((unsigned char *) p) = c; // shadow

        LOOP_TAIL_X
      }
    }
  }
}

#undef LOOP_HEAD
#undef LOOP_TAIL
#undef LOOP_TAIL_X

File: txtscrio/pas/ShowCStr.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure ShowCStr(dest: tSCREEN_MEM_PTR; x,y: Byte; str: String; atr1,atr2: Byte);

var
  s: ^Char; // input string
  p: ^Word; // output screen address
  a1,a2,t: Word; // prepared attributes, `t' for swapping
  n: Byte; // number of input characters left to proceed
  c: Char; // character to show
  flag: Byte; // 0=read, 1=stop, 2=show

begin
  n := Length (str);
  if (n <> 0) then begin
    s := @str[1];
    PosChar (x, y); // set `absolute_pos'
    p := Pointer(dest) + absolute_pos; // cast to `Pointer' for correct address
    a1 := atr1 shl 8;
    a2 := atr2 shl 8;
    repeat
      repeat
        c := s^;
        Inc (s);

        if (c = '~') then begin // color switch mark
          t := a2;
          a2 := a1;
          a1 := t;
          Dec (n);
          flag := Ord(n = 0); // check string end
        end else
          flag := 2;

      until (flag <> 0); // color switch mark

      if (flag = 1) then break;

      p^ := Ord(c) + a1;
      Inc (p); // in 2-bytes units

      Dec (n);
    until (n = 0); // check string end
  end;
end;
